src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -43,6 +45,153 @@ void Bootstrap(const MatType& dataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    mlpack::Bootstrap<UseWeights>(dataset,
+                                  labels,
+                                  weights,
+                                  bootstrapDataset,
+                                  bootstrapLabels,
+                                  bootstrapWeights);
+  }
+};
+
+/**
+ * No-op bootstrap strategy that just copies the input to the output.
+ */
+template<bool UseWeights>
+class IdentityBootstrap
+{
+ public:
+  template<typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    bootstrapDataset = dataset;
+    bootstrapLabels = labels;
+    if (UseWeights)
+      bootstrapWeights = weights;
+  }
+};
+
+/**
+ * Sequential bootstrap.
+ */
+template<bool UseWeights, typename G = std::mt19937>
+class SequentialBootstrap
+{
+ public:
+  explicit SequentialBootstrap(const arma::mat& indicatorMatrix,
+                               const G& generator = G(std::random_device()())) :
+    SequentialBootstrap(indicatorMatrix, indicatorMatrix.n_rows, generator)
+  {
+  }
+
+  explicit SequentialBootstrap(const arma::mat& indicatorMatrix,
+                               arma::uword sampleCount,
+                               const G& generator = G(std::random_device()())):
+    indicatorMatrix(indicatorMatrix),
+    colIndices(arma::linspace<arma::uvec>(0, indicatorMatrix.n_cols - 1)),
+    sampleCount(sampleCount),
+    generator(generator)
+  {
+  }
+
+  template<typename MatType,
+           typename LabelsType,
+           typename WeightsType>
+  void Bootstrap(const MatType& dataset,
+                 const LabelsType& labels,
+                 const WeightsType& weights,
+                 MatType& bootstrapDataset,
+                 LabelsType& bootstrapLabels,
+                 WeightsType& bootstrapWeights)
+  {
+    if (labels.n_cols != indicatorMatrix.n_rows)
+      throw std::invalid_argument("SequentialBootstrap::Bootstrap(): "
+          "labels and indicatorMatrix n_rows differ!");
+
+    // observations are stored as columns and dimensions
+    // (number of features) as rows.
+
+    const arma::uvec phi(ComputeSamples());
+
+    bootstrapDataset = dataset.cols(phi);
+    bootstrapLabels = labels.cols(phi);
+    if (UseWeights)
+      bootstrapWeights = weights.rows(phi);
+  }
+
+  arma::uvec ComputeSamples() const
+  {
+    arma::uvec phi;
+
+    while (phi.size() < sampleCount) {
+      const auto prob(ComputeNextDrawProbabilities(phi, indicatorMatrix));
+      std::discrete_distribution dist(prob.begin(), prob.end());
+
+      phi = arma::join_cols(phi, arma::uvec{ colIndices[dist(generator)] });
```

Comment:
This will actually be quite slow: it will reallocate `phi` every iteration.  I understand that the algorithm will use a different number of samples for each bootstrap set.  It may be better to use something like `std::vector<size_t>` here instead to hold the indices, since amortized performance of `.push_back()` is O(1) (if I remember right).  Basically it over-allocates when you add elements, and this significantly reduces the number of allocations.  But Armadillo will not do this.  Once the `std::vector<>` is fully built, you can convert directly to an `arma::uvec` with `conv_to`.